### PR TITLE
Support general SIMD instruction

### PR DESCRIPTION
**Pull request by harihitode**
Making ndt_omp independent to the SIMD architecture

* using compiler flag -march=native, instead of using sse flags immediately.
* using aligned allocation std::vector of Eigen::Matrix

### CODE REVIEW

**Comment by kenji-miyake** [comment]
I'll close and re-open since I changed the target branch to try to run CI workflows.
-> But it seems there is no effect. :cry:

**Review by KeisukeShima — APPROVED** [review]

**Comment by KeisukeShima** [comment]
@kenji-miyake Please double-check this PR :pray:

File: CMakeLists.txt
Comment (kenji-miyake) [inline CMakeLists.txt:6]: Will removing `-msse -msse2 -msse3 -msse4 -msse4.1 -msse4.2` not affect the performance?
Could you link any references around these options?

Also, are there any reasons for removing `-std=c++14`?

File: include/pclomp/ndt_omp.h
Comment (kenji-miyake) [inline include/pclomp/ndt_omp.h:231]: Could you link Eigen's documents as well?
Since I think we'll submit a PR to upstream, we need some references to explain to the maintainer.

**Comment by kenji-miyake** [comment]
Also, why the CI has failed? I think we should fix them.

File: CMakeLists.txt
Comment (KeisukeShima) [reply CMakeLists.txt:6]: > Will removing -msse -msse2 -msse3 -msse4 -msse4.1 -msse4.2 not affect the performance?

Yes, @harihitode san evaluated the performance.
https://tier4.atlassian.net/browse/T4PB-12676?focusedCommentId=69870

Here is the reference.
http://eigen.tuxfamily.org/index.php?title=FAQ#How_can_I_enable_vectorization.3F

> Also, are there any reasons for removing -std=c++14?

Ah, I guess we need to keep this option. It's better to use `set(CMAKE_CXX_STANDARD 14)` instead.

File: include/pclomp/ndt_omp.h
Comment (KeisukeShima) [reply include/pclomp/ndt_omp.h:231]: I guess it has something to do with this issue https://github.com/isl-org/Open3D/issues/653

File: CMakeLists.txt
Comment (kenji-miyake) [reply CMakeLists.txt:6]: > Yes, @harihitode san evaluated the performance.
https://tier4.atlassian.net/browse/T4PB-12676?focusedCommentId=69870

Since it's an internal link, we should copy the result here so that everyone can see it.

> Ah, I guess we need to keep this option. It's better to use set(CMAKE_CXX_STANDARD 14) instead.

Yes, but making it another PR is better.

Comment (KeisukeShima) [reply CMakeLists.txt:6]: > Yes, but making it another PR is better.

OK. Then I restore c++14 option.
3b3b92a

Comment (KeisukeShima) [reply CMakeLists.txt:6]: > Since it's an internal link, we should copy the result here so that everyone can see it.

image

Comment (kenji-miyake) [reply CMakeLists.txt:6]: Thank you, but you should write the condition of the experiment.

**Comment by KeisukeShima** [comment]
The CI target is noetic and melodic. CI scripts for ROS2 is ready for review on #5

**Comment by kenji-miyake** [comment]
Let's merge this after #5 is merged and the CI is checked in this PR.

@KeisukeShima @harihitode Could you send PRs equivalent with #5 and #8 to https://github.com/koide3/ndt_omp?

**Comment by kenji-miyake** [comment]
@KeisukeShima Could you rebase this, please? :pray:

**Comment by KeisukeShima** [comment]
This repository does not allow force push, so I created a new PR. #9

**Comment by kenji-miyake** [comment]
hmm, actually it was possible.

**Review by kenji-miyake — APPROVED** [review]
@yukkysaito @YamatoAndo @e-takeuchi I'd like to merge this PR for ARM support.
According to the evaluation result by @harihitode, there is no performance degradation.

Is that okay for you?

File: CMakeLists.txt
Comment (kenji-miyake) [inline CMakeLists.txt:5]: I think you don't have to write the same options twice.

```
add_definitions(-std=c++14)
set(CMAKE_CXX_FLAGS "-std=c++14")
```

Comment (yukkysaito) [reply CMakeLists.txt:5]: If this compile option is changed, will it affect SIMD?

Comment (kenji-miyake) [reply CMakeLists.txt:5]: @yukkysaito Thank you! Is it the same comment as this?
https://github.com/tier4/ndt_omp/pull/8#discussion_r783663368

Or do you mean `std=c++14`? (I assume not.)

Comment (yukkysaito) [reply CMakeLists.txt:5]: @kenji-miyake 

>Is it the same comment as this?

Ops! it is same thing. :pray:

**Comment by yukkysaito** [comment]
@kenji-miyake It is good, although there is a question of whether the same results can be obtained on real machines on vehicles.

File: CMakeLists.txt
Comment (harihitode) [reply CMakeLists.txt:6]: @kenji-miyake @KeisukeShima The result is obtained on Intel machine using rosbag playing simulation.

CPU: i9-9900 supporting AVX256 (256 bit SIMD)
We play ROSBAG as instructed in https://tier4.github.io/autoware.proj/tree/main/tutorial/QuickStart/ .

Red line is our proposal which uses 256 bit SIMD in the execution binary. Current default is the blue line, 128 bit SIMD.
Because this program itself is not tuned to the 256 bit SIMD, we have no performance-up either performance degradation.

Comment (harihitode) [reply CMakeLists.txt:6]: @yukkysaito Thanks for your comment. I think this change has no effect on its performance, but I have evaluated it only on rosbag-simulation.

File: CMakeLists.txt
Comment (KeisukeShima) [reply CMakeLists.txt:5]: > I think you don't have to write the same options twice.
> 
> ```
> add_definitions(-std=c++14)
> set(CMAKE_CXX_FLAGS "-std=c++14")
> ```

In the previous comment, you said it is good to create another PR to modify it.
https://github.com/tier4/ndt_omp/pull/8#discussion_r783672936

Comment (kenji-miyake) [reply CMakeLists.txt:5]: I see, it was written twice in the original code as well. I'm sorry I overlooked it.

**Review by yukkysaito — APPROVED** [review]

**Pull request merged**
